### PR TITLE
Inference Compilation example

### PR DESCRIPTION
**Pull request by wsgharvey**
An example using inference compilation. Requires pull request #1439 to run.

### CODE REVIEW

**Comment by eb8680** [comment]
Thanks for adding this! Since this is a notebook, can you move it to `tutorials/source/` and add a link to it in `tutorials/source/index.rst`?

**Comment by martinjankowiak** [comment]
thanks @wsgharvey this is great. could you maybe try to add a bit more context/hand-holding to make this more useful to a broader audience? in particular a link to/description of the compilation approach would be super useful. 

of course, if you don't have the time, we can always refine/expand the text a bit in a subsequent PR.

**Comment by wsgharvey** [comment]
Sure, I'll try to expand it a bit later today

File: tutorial/source/csis.ipynb [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+   "metadata": {},
+   "source": [
+    "Using $x_1 = 8$ and $x_2 = 9$ makes inference difficult (both for CSIS and regular importance sampling), since pairs of observations greater than this will only be encountered roughly once every 1500 samples, and so will have been encountered by the guide only once or twice during compilation."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### We now plot the results and compare with importance sampling:"
```
Comment (jpchen) [inline tutorial/source/csis.ipynb]: a few lines of explanation interpreting the results would be helpful

File: tutorial/source/index.rst [hunk 2]
```diff
@@ -35,6 +35,7 @@ Welcome to Pyro Examples and Tutorials!
    gp
    bo
    tracking_1d
+   csis_gaussian_unknown_mean
```
[the remark this answers is not in the document]
Comment (wsgharvey) [reply tutorial/source/index.rst]: Good point, I'd forgotten to change that

File: tutorial/source/csis.ipynb [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+   "source": [
+    "# Compiled Sequential Importance Sampling\n",
+    "\n",
+    "Compiled sequential importance sampling [1], or inference compilation, is a technique to amortize the computational cost of inference by learning a proposal distribution for importance sampling.\n",
+    "\n",
+    "The proposal distribution is learned to minimise the KL divergence between the model and the guide, $ \\rm{KL}\\!\\left( p({\\bf z} | {\\bf x}) \\lVert q_{\\phi, x}({\\bf z}) \\right) $. This differs from variational inference, which would minimise $ \\rm{KL}\\!\\left( q_{\\phi, x}({\\bf z}) \\lVert p({\\bf z} | {\\bf x}) \\right) $. Using this loss encourages the approximate proposal distribution to be broader than the true posterior (mass covering), whereas variational inference typically learns a narrower approximation (mode seeking). Guides for importance sampling are usually desired to have heavier tails than the model (see <https://stats.stackexchange.com/questions/76798/in-importance-sampling-why-should-the-importance-density-have-heavier-tails>). Therefore, the inference compilation loss is usually more suited to compiling a guide for importance sampling.\n",
```
[the remark this answers is not in the document]
Comment (wsgharvey) [reply tutorial/source/csis.ipynb]: Yep, that would be nicer

**Pull request merged**